### PR TITLE
Allow caller to handle help/argument exceptions

### PR DESCRIPTION
**Pull request by mtmcp**
This change improves robustness of argument-parsing logic by allowing the caller to decide what to do when `-h/--help` is supplied, or when an invalid argument is supplied by the user.

### CODE REVIEW

**Comment by cebtenzzre** [comment]
Is the only functional change in this 100-line PR the fact that log_print_usage is now also called on errors? What would the caller want to do besides end the program if invalid command-line arguments were passed?

**Comment by bandoti** [comment]
> Is the only functional change in this 100-line PR the fact that log_print_usage is now also called on errors? What would the caller want to do besides end the program if invalid command-line arguments were passed?

Currently, applications in the examples directory simply print the usage message. However, when embedding the common logic as a C/C++ plugin for interpreted languages (which is something I'm working on), it is _always_ desirable *not* to exit when the arguments are wrong or help is requested. The desired behavior in that case is to let the higher-level language determine if/when to exit the application.

Due to the complexity and sheer number of arguments which can be supplied to llama-based applications, it is desirable to have the common logic parse/process command-line arguments instead of having to duplicate the work.

**Comment by cebtenzzre** [comment]
Are you proposing that common.cpp should become a public API? As far as I know, it is only intended to be a place for code that is shared between the internal example programs.

**Comment by bandoti** [comment]
> Are you proposing that common.cpp should become a public API? As far as I know, it is only intended to be a place for code that is shared between the internal example programs.

I think there is sufficient functionality within common to consider moving (at least some) to a public API. As is evident by the number of applications within `examples` which use the functionality, there is certainly a use-case. The need, in my opinion, is to provide a "gentler" API for folks (mostly everyone?) who don't want to reinvent the wheel.

While that is a lot broader in scope than the changes proposed in this PR—a more flexible way to respond to user input always helps toward that end. 🙂

**Comment by bandoti** [comment]
@cebtenzzre Per your feedback I reduced the scope of the change by hiding the implementation in a new `gpt_params_parse_ex` function. That way, the code change is reduced to only `common.h/cpp`.

File: common/common.h [hunk 1]
```diff
@@ -110,6 +110,8 @@ struct gpt_params {
     std::string image = ""; // path to an image file
 };
 
+bool gpt_params_parse_ex(int argc, char** argv, gpt_params& params);
```
Comment (ggerganov) [inline common/common.h:113]:
```suggestion
bool gpt_params_parse_ex(int argc, char ** argv, gpt_params & params);
```

File: common/common.cpp [hunk 2]
```diff
@@ -102,10 +102,25 @@ void process_escapes(std::string& input) {
     input.resize(output_idx);
 }
 
-bool gpt_params_parse(int argc, char ** argv, gpt_params & params) {
+bool gpt_params_parse(int argc, char** argv, gpt_params& params) {
```
Comment (ggerganov) [inline common/common.cpp:105]:
```suggestion
bool gpt_params_parse(int argc, char ** argv, gpt_params & params) {
```

**Review by ggerganov — APPROVED** [review]
The common code might become some sort of higher-level API in the future, but currently this is not the focus and there are not plans to do so. It's strongly recommended that 3rd party projects avoid using it or depending on it.

**Pull request merged**